src/Microsoft.AspNetCore.Http.Connections.Client/HttpConnection.cs

```diff
@@ -62,7 +62,14 @@ public override IDuplexPipe Transport
         }
 
         public override IFeatureCollection Features { get; } = new FeatureCollection();
-        public override string ConnectionId { get; set; }
+
+        // ConnectionId will be null if negotate was skipped
+        public override string ConnectionId
+        {
+            get => _connectionId;
+            set => throw new InvalidOperationException("The ConnectionId is set internally when the HttpConnection starts by negotate.");
```

Comment:
"The ConnectionId is set internally and should not be set by user code" (no need to go into detail about how it's set :))

Reply:
This being our last chance to do so, should we consider making `ConnectionContext.ConnectionId` read-only? We generally make these properties read-write for testability (following in `HttpContext`s footsteps), but this is a bit of a special case.

Reply:
It's get/set on `ConnectionContext`.

Reply:
> (no need to go into detail about how it's set :))

Devs might wonder about when it is set. Also as far as I can tell it is only set when there is a negotiate. If skipNegotiate is true then it will be null.

Reply:
No it shouldn’t be readonly. I’m not sure about throwing either

Reply:
> If skipNegotiate is true then it will be null.

That's correct but it's "by-design" in that no connection ID is needed when using a direct WebSocket connection (since it's handled by a single request)

Reply:
Yep!